File: c/programs/sum-four-fifth-powers-is-perfect-fitfh-power-optimized.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

typedef struct {
    unsigned long long sum;
    int c;
    int d;
} CDSum;

typedef struct {
    int a, b, c, d, e;
} Solution;

int compare(const void *a, const void *b) {
    const CDSum *sa = (const CDSum *)a;
    const CDSum *sb = (const CDSum *)b;
    return (sa->sum > sb->sum) - (sa->sum < sb->sum);
}

void find_solutions(int upper_limit, Solution **solutions, int *count) {
    *count = 0;
    *solutions = NULL;
    
    // Store FIRST valid (c,d) for each sum
    CDSum *sum_cd = malloc(upper_limit * (upper_limit + 1) / 2 * sizeof(CDSum));
    int index = 0;
    
    for (int c = 1; c <= upper_limit; c++) {
        unsigned long long c5 = (unsigned long long)c * c * c * c * c;
        for (int d = 1; d <= c; d++) {
            unsigned long long s = c5 + (unsigned long long)d * d * d * d * d;
            // Only store first occurrence of each sum
            if (index == 0 || s > sum_cd[index-1].sum) {
                sum_cd[index].sum = s;
                sum_cd[index].c = c;
                sum_cd[index].d = d;
                index++;
            }
        }
    }

    // Search phase remains O(n³)
    for (int a = 1; a <= upper_limit; a++) {
        unsigned long long a5 = (unsigned long long)a * a * a * a * a;
        for (int b = 1; b <= a; b++) {
            unsigned long long b5 = (unsigned long long)b * b * b * b * b;
            unsigned long long s_ab = a5 + b5;
            int e_max = (int)(pow(s_ab + 2*b5, 0.2)) + 2;

            for (int e = 1; e <= e_max; e++) {
                unsigned long long target = (unsigned long long)e * e * e * e * e - s_ab;
                CDSum key = {target, 0, 0};
                CDSum *found = bsearch(&key, sum_cd, index, sizeof(CDSum), compare);
                if (found && found->c <= b && found->d <= found->c) {
                    *solutions = realloc(*solutions, (*count + 1) * sizeof(Solution));
                    (*solutions)[*count] = (Solution){a, b, found->c, found->d, e};
                    (*count)++;
                }
            }
        }
    }
    free(sum_cd);
}

File: c/programs/sum-four-fifth-powers-is-perfect-fitfh-power-optimized.py
from collections import defaultdict
import time


def find_solutions(upper_limit):
    # Store FIRST valid (c,d) for each sum
    sum_cd = {}
    for c in range(1, upper_limit + 1):
        c5 = c**5
        for d in range(1, c + 1):
            s = c5 + d**5
            if s not in sum_cd:  # Store only first occurrence
                sum_cd[s] = (c, d)

    solutions = []
    for a in range(1, upper_limit + 1):
        a5 = a**5
        for b in range(1, a + 1):
            s_ab = a5 + b**5
            e_max = int((s_ab + 2*b**5)**0.2) + 2
            for e in range(1, e_max + 1):
                target = e**5 - s_ab
                if target in sum_cd:
                    c, d = sum_cd[target]
                    if c <= b and d <= c:
                        solutions.append((a, b, c, d, e))
    return solutions
